Aggregate cost metric

kathy-t opened the pull request:
**Description**
This PR aggregates the cost metric.

Companion PRs:
- https://github.com/dockstore/dockstore/pull/5577
- https://github.com/dockstore/dockstore-ui2/pull/1805

**Review Instructions**
- Submit a run execution metric that specifies a cost for a workflow. Read this [doc](https://docs.dockstore.org/en/stable/advanced-topics/metrics.html#how-do-i-submit-workflow-metrics) for more info if you're unfamiliar with submitting metrics. You must be a curator or admin to be able to submit metrics.
- Perform a release of the metrics aggregator so that the new JAR is in artifactory. View the [wiki](https://wiki.oicr.on.ca/display/DOC/Dockstore+Releases#DockstoreReleases-RepositoriesthatuseFriendlyCIversionsandaGitHubAction) for how to do this
  - [ ] Note to self: release after this is merged so reviewing is easier for the verifier
 - SSM onto the deployer and wget the metrics aggregator JAR from artifactory
 - Aggregate the metrics. View the [README](https://github.com/dockstore/dockstore-support/tree/develop/metricsaggregator#metrics-aggregator) for instructions.
 - Go to the workflow version that you submitted metrics for in the UI and enable the metrics flag. Verify that the cost metric shows up.

**Issue**
[SEAB-5625](https://ucsc-cgl.atlassian.net/browse/SEAB-5625)

**Security**
If there are any concerns that require extra attention from the security team, highlight them here.

Please make sure that you've checked the following before submitting your pull request. Thanks!

- [x] Check that you pass the basic style checks and unit tests by running `mvn clean install` in the project that you have modified (until https://ucsc-cgl.atlassian.net/browse/SEAB-5300 adds multi-module support properly)
- [x] Ensure that the PR targets the correct branch. Check the milestone or fix version of the ticket.
- [x] If you are changing dependencies, check with dependabot to ensure you are not introducing new high/critical vulnerabilities
- [x] If this PR is for a user-facing feature, create and link a documentation ticket for this feature (usually in the same milestone as the linked issue). Style points if you create a documentation PR directly and link that instead. 


[SEAB-5625]: https://ucsc-cgl.atlassian.net/browse/SEAB-5625?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

coverbeck commented on metricsaggregator/src/main/java/io/dockstore/metricsaggregator/helper/AggregationHelper.java:
```diff
@@ -259,6 +261,53 @@ public static Optional<MemoryMetric> getAggregatedMemoryFromExecutions(List<RunE
+        List<CostMetric> costMetrics = allSubmissions.getAggregatedExecutions().stream()
+                .map(Metrics::getCost)
+                .filter(Objects::nonNull)
+                .collect(Collectors.toList());
```
You're using this instead of `toList` because you need a mutable list for the next line, right? If so, the `Collectors.toList()` JavaDoc says "There are no guarantees on the type, mutability, serializability, or thread-safety of the {@code List} returned". The current implementation returns an ArrayList, so it's good, but in theory it could break in a future JDK release (but unlikely).

kathy-t replied:
Yeah I need a mutable list. Would wrapping it in `ArrayList` be the way to go?

```
List<CostMetric> costMetrics = new ArrayList<>(allSubmissions.getAggregatedExecutions().stream()
                .map(Metrics::getCost)
                .filter(Objects::nonNull)
                .toList());
```

svonworl replied:
That works, you can also:
```
.collect(Collectors.toCollection(ArrayList::new))
```

kathy-t commented:
Changes:
- Based on feedback in https://github.com/dockstore/dockstore/pull/5577#discussion_r1245671082, changed the aggregation of the cost metrics to use the Java Money library to preserve accuracy. In doing so, I removed the `Statistics` record and made it an abstract class that is inherited by `DoubleStatistics` and `MoneyStatistics`. I thought this was a better approach than adding a bunch of Money specific functions to the `Statistics` record

denis-yuen commented on THIRD-PARTY-LICENSES.txt:
```diff
@@ -124,10 +124,10 @@ Lists of 399 third-party dependencies.
-     (Apache Software License, Version 2.0) dockstore-integration-testing (io.dockstore:dockstore-integration-testing:1.15.0-alpha.4 - no url defined)
-     (Apache Software License, Version 2.0) dockstore-language-plugin-parent (io.dockstore:dockstore-language-plugin-parent:1.15.0-alpha.4 - no url defined)
-     (Apache Software License, Version 2.0) dockstore-webservice (io.dockstore:dockstore-webservice:1.15.0-alpha.4 - no url defined)
+     (Apache Software License, Version 2.0) dockstore-common (io.dockstore:dockstore-common:1.15.0-SNAPSHOT - no url defined)
```
Remember to change this (snapshot)

denis-yuen commented on metricsaggregator/src/main/java/io/dockstore/metricsaggregator/DoubleStatistics.java:
```diff
@@ -0,0 +1,114 @@
+     */
+    @Override
+    public Double calculateAverage(List<Double> dataPoints) {
+        return dataPoints.stream().mapToDouble(d -> d).average().getAsDouble();
```
Would have thought averaging would be one of the dangerous operations requiring `BigDecimal` or `Money`
https://stackoverflow.com/questions/54278847/averaging-bigdecimals-using-streams-api-collectors

denis-yuen commented on metricsaggregator/src/main/java/io/dockstore/metricsaggregator/DoubleStatistics.java:
```diff
@@ -0,0 +1,114 @@
+        int totalNumberOfDataPoints = getTotalNumberOfDataPoints(statistics);
+        return statistics.stream()
+                .map(stat -> {
+                    double weight = (double)stat.getNumberOfDataPoints() / (double)totalNumberOfDataPoints;
```
Ditto

denis-yuen commented on metricsaggregator/src/main/java/io/dockstore/metricsaggregator/MoneyStatistics.java:
```diff
@@ -0,0 +1,90 @@
+        int totalNumberOfDataPoints = getTotalNumberOfDataPoints(statistics);
+        return statistics.stream()
+                .map(stat -> {
+                    double weight = (double)stat.getNumberOfDataPoints() / (double)totalNumberOfDataPoints;
```
Ok, nvm, the above. 
For my understanding, we still retain a "normal" `double` class for statistics where precision is less important and have a class using `Money` (which seems ok).

That said, I think averaging would be one operation which would be problematic (includes division and multiplying)

kathy-t replied:
> For my understanding, we still retain a "normal" double class for statistics where precision is less important and have a class using Money (which seems ok).
That said, I think averaging would be one operation which would be problematic (includes division and multiplying)

Should I change the [webservice](https://github.com/dockstore/dockstore/blob/8da78281c6881a4a543f1c2a84085867f5f272d7/dockstore-webservice/src/main/java/io/dockstore/webservice/core/metrics/StatisticMetric.java#L65) `StatisticMetric` fields to use `BigDecimal` instead of `double` then? This change would ripple to the metrics aggregator

Double checking, are you against the use of `double` to calculate the weight (on the line you commented on)? Should it be a `BigDecimal` instead?

kathy-t replied:
> That said, I think averaging would be one operation which would be problematic (includes division and multiplying)

This averaging is done with the Java Money class, i.e. `stat.getAverage()` is working with a `Money` object, not a double. The double is only used to calculate the weight.

coverbeck commented on metricsaggregator/pom.xml:
```diff
@@ -131,6 +131,14 @@
+            <artifactId>moneta-core</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>javax.money</groupId>
```
I did not this existed. :)

kathy-t commented on metricsaggregator/src/main/java/io/dockstore/metricsaggregator/helper/AggregationHelper.java:
```diff
@@ -249,12 +258,71 @@ public static Optional<MemoryMetric> getAggregatedMemoryFromExecutions(List<RunE
+                .filter(Objects::nonNull)
+                .toList();
+
+        boolean containsMalformedCurrencies = submittedCosts.stream().anyMatch(cost -> !isValidCurrencyCode(cost.getCurrency()));
```
The Java Money library can [convert currencies](https://www.baeldung.com/java-money-and-currency#currency-conversion), so I think we would convert it all to one currency (maybe USD) then aggregate

coverbeck replied:
Interesting, I did not know that. 

I don't know finance, but I wonder if we should track when we calculate the exchange? For example, a currency is devalued; if we calculate the exchange the day before the devaluation, it evaluates to $10, but if we calculate it the day after the devaluation, it comes to $6...

But I don't know finance, and we're getting into real hypotheticals here; I don't think this needs to block the PR.

kathy-t replied:
That's a good point and something I did ponder when deciding if the endpoint should allow the submission of any currency in the first iteration of the cost metric. Ultimately decided to keep things simple by only allowing USD submissions, so we don't need to worry about things like that

This does make me wonder if we should record the date that the workflow was executed because a workflow run that costs $5 in 2020 might not cost $5 today :thinking:

denis-yuen reviewed (approved):
Very close, the SonarCloud bug warnings seem worth looking into (basically check whether an optional exists before using it)
